File: sdk/python/src/metergraph/_transport.py
"""Bounded, fire-and-forget HTTP transport using only the stdlib."""

from __future__ import annotations

import gzip
import json
import logging
import os
import queue
import threading
import time
import urllib.error
import urllib.request
from typing import Any


log = logging.getLogger("metergraph")
MAX_BATCH_BYTES = 512 * 1024


class Writer:
    def __init__(
        self,
        token: str,
        base_url: str,
        *,
        queue_size: int = 2000,
        batch_size: int = 100,
        flush_seconds: float = 5.0,
    ) -> None:
        self._token = token
        self._url = f"{base_url.rstrip('/')}/v1/ingest"
        self._queue_size = max(1, queue_size)
        self._batch_size = max(1, min(batch_size, 1000))
        self._flush_seconds = max(0.05, flush_seconds)
        self._queue: queue.Queue[dict[str, Any]] = queue.Queue(self._queue_size)
        self._stop = threading.Event()
        self._flush_now = threading.Event()
        self._fatal = False
        self._dropped = 0
        self._errors = 0
        self._backoff = 1.0
        self._retry_at = 0.0
        self._thread = self._new_thread()
        self._thread.start()
        if hasattr(os, "register_at_fork"):
            os.register_at_fork(after_in_child=self._after_fork)

    @property
    def dropped(self) -> int:
        return self._dropped

    def _new_thread(self) -> threading.Thread:
        return threading.Thread(target=self._run, name="metergraph-writer", daemon=True)

    def _after_fork(self) -> None:
        # Inherited Queue locks and threads are not safe in the child. Pending
        # parent rows stay with the parent; the child starts a clean writer.
        self._queue = queue.Queue(self._queue_size)
        self._stop = threading.Event()
        self._flush_now = threading.Event()
        self._thread = self._new_thread()
        self._thread.start()

    def enqueue(self, row: dict[str, Any]) -> bool:
        if self._fatal or self._stop.is_set():
            self._dropped += 1
            return False
        try:
            self._queue.put_nowait(row)
            return True
        except queue.Full:
            self._dropped += 1
            return False

    def _run(self) -> None:
        while not self._stop.is_set() or not self._queue.empty():
            try:
                first = self._queue.get(timeout=min(self._flush_seconds, 0.1))
            except queue.Empty:
                continue
            batch = [first]
            deadline = time.monotonic() + self._flush_seconds
            while len(batch) < self._batch_size:
                if self._flush_now.is_set():
                    break
                try:
                    batch.append(self._queue.get_nowait())
                except queue.Empty:
                    remaining = deadline - time.monotonic()
                    if remaining <= 0 or self._stop.is_set():
                        break
                    try:
                        batch.append(self._queue.get(timeout=min(remaining, 0.05)))
                    except queue.Empty:
                        continue
            try:
                self._deliver(batch)
            finally:
                self._flush_now.clear()
                for _ in batch:
                    self._queue.task_done()

    def _deliver(self, rows: list[dict[str, Any]]) -> bool:
        if self._fatal or time.monotonic() < self._retry_at:
            self._dropped += len(rows)
            return False
        meta = {"dropped": self._dropped, "transport_errors": self._errors}
        body = json.dumps(
            {"schema_version": 1, "rows": rows, "meta": meta},
            separators=(",", ":"),
        ).encode()
        headers = {
            "Authorization": f"Bearer {self._token}",
            "Content-Type": "application/json",
            "User-Agent": "metergraph-python/0.1.0",
        }
        if len(body) > 32 * 1024:
            body = gzip.compress(body)
            headers["Content-Encoding"] = "gzip"
        if len(body) > MAX_BATCH_BYTES:
            if len(rows) == 1:
                self._dropped += 1
                return False
            midpoint = len(rows) // 2
            left = self._deliver(rows[:midpoint])
            right = self._deliver(rows[midpoint:])
            return left and right
        request = urllib.request.Request(
            self._url, data=body, headers=headers, method="POST"
        )
        try:
            with urllib.request.urlopen(request, timeout=5) as response:
                if response.status != 202:
                    raise OSError(f"unexpected ingest status {response.status}")
            self._backoff = 1.0
            self._retry_at = 0.0
            return True
        except urllib.error.HTTPError as exc:
            if exc.code in (401, 403):
                self._fatal = True
                log.warning(
                    "Metergraph authentication failed; capture disabled for this process"
                )
            else:
                self._failed(len(rows))
            return False
        except Exception:
            self._failed(len(rows))
            return False

    def _failed(self, count: int) -> None:
        self._errors += 1
        self._dropped += count
        self._retry_at = time.monotonic() + self._backoff
        self._backoff = min(self._backoff * 2, 60.0)

    def flush(self, timeout: float = 3.0) -> bool:
        deadline = time.monotonic() + max(0.0, timeout)
        self._flush_now.set()
        while self._queue.unfinished_tasks and time.monotonic() < deadline:
            time.sleep(0.01)
        return self._queue.unfinished_tasks == 0

    def shutdown(self) -> None:
        self.flush(3.0)
        self._stop.set()
        if threading.current_thread() is not self._thread:
            self._thread.join(timeout=max(1.0, self._flush_seconds + 0.5))

File: sdk/typescript/src/transport.ts
export type TransportMode = "auto" | "background" | "buffered";
export type WaitUntil = (promise: Promise<unknown>) => void;
export const MAX_BATCH_BYTES = 512 * 1024;

export interface TransportOptions {
  queueSize?: number;
  batchSize?: number;
  flushMs?: number;
  mode?: TransportMode;
}

function detectedMode(): Exclude<TransportMode, "auto"> {
  const env = typeof process === "undefined" ? {} : process.env;
  const agent = typeof navigator === "undefined" ? "" : navigator.userAgent;
  return env?.AWS_LAMBDA_FUNCTION_NAME || env?.VERCEL || agent === "Cloudflare-Workers"
    ? "buffered"
    : "background";
}

async function gzipBody(body: ArrayBuffer): Promise<ArrayBuffer> {
  if (typeof CompressionStream === "undefined") return body;
  const source = new Blob([body]).stream();
  return new Response(source.pipeThrough(new CompressionStream("gzip"))).arrayBuffer();
}

export class Transport {
  private readonly queue: Record<string, unknown>[] = [];
  private readonly mode: Exclude<TransportMode, "auto">;
  private readonly queueSize: number;
  private readonly batchSize: number;
  private timer?: ReturnType<typeof setInterval>;
  private waitUntil?: WaitUntil;
  private scheduled = false;
  private inFlight: Promise<void> = Promise.resolve();
  private fatal = false;
  private dropped = 0;
  private errors = 0;
  private retryAt = 0;
  private backoffMs = 1_000;

  constructor(
    private readonly token: string,
    private readonly baseUrl: string,
    options: TransportOptions = {},
  ) {
    this.queueSize = Math.max(1, options.queueSize ?? 2_000);
    this.batchSize = Math.max(1, Math.min(1_000, options.batchSize ?? 100));
    this.mode = options.mode === "auto" || !options.mode ? detectedMode() : options.mode;
    if (this.mode === "background") {
      this.timer = setInterval(() => void this.flush(), Math.max(50, options.flushMs ?? 5_000));
      this.timer.unref?.();
    }
  }

  bindWaitUntil(value: { waitUntil(promise: Promise<unknown>): void } | WaitUntil): void {
    this.waitUntil = typeof value === "function" ? value : value.waitUntil.bind(value);
  }

  enqueue(row: Record<string, unknown>): boolean {
    if (this.fatal || this.queue.length >= this.queueSize) {
      this.dropped += 1;
      return false;
    }
    this.queue.push(row);
    if (this.queue.length >= this.batchSize) this.schedule();
    else if (this.mode === "buffered") this.schedule();
    return true;
  }

  private schedule(): void {
    if (this.scheduled) return;
    this.scheduled = true;
    const work = Promise.resolve().then(async () => {
      this.scheduled = false;
      await this.flush();
    });
    if (this.waitUntil) this.waitUntil(work);
    else if (this.mode === "buffered") void work;
  }

  async flush(timeoutMs = 3_000): Promise<boolean> {
    const deadline = Date.now() + Math.max(0, timeoutMs);
    const work = async () => {
      while (this.queue.length && Date.now() < deadline) {
        const batch = this.queue.splice(0, this.batchSize);
        await this.deliver(batch);
      }
    };
    this.inFlight = this.inFlight.then(work, work);
    await new Promise<void>((resolve) => {
      const timer = setTimeout(resolve, Math.max(0, deadline - Date.now()));
      timer.unref?.();
      void this.inFlight.finally(() => {
        clearTimeout(timer);
        resolve();
      });
    });
    return this.queue.length === 0;
  }

  private async deliver(rows: Record<string, unknown>[]): Promise<void> {
    if (this.fatal || Date.now() < this.retryAt) {
      this.dropped += rows.length;
      return;
    }
    const encoded = new TextEncoder().encode(JSON.stringify({
      schema_version: 1,
      rows,
      meta: { dropped: this.dropped, transport_errors: this.errors },
    }));
    const raw = encoded.buffer.slice(
      encoded.byteOffset, encoded.byteOffset + encoded.byteLength,
    ) as ArrayBuffer;
    const compressed = raw.byteLength > 32 * 1024;
    const body = compressed ? await gzipBody(raw) : raw;
    if (body.byteLength > MAX_BATCH_BYTES) {
      if (rows.length === 1) {
        this.dropped += 1;
        return;
      }
      const midpoint = Math.floor(rows.length / 2);
      await this.deliver(rows.slice(0, midpoint));
      await this.deliver(rows.slice(midpoint));
      return;
    }
    try {
      const response = await fetch(`${this.baseUrl.replace(/\/$/, "")}/v1/ingest`, {
        method: "POST",
        headers: {
          Authorization: `Bearer ${this.token}`,
          "Content-Type": "application/json",
          ...(compressed && body !== raw ? { "Content-Encoding": "gzip" } : {}),
        },
        body: new Blob([body], { type: "application/json" }),
        keepalive: this.mode === "buffered" && body.byteLength <= 64 * 1024,
      });
      if (response.status === 401 || response.status === 403) {
        this.fatal = true;
        console.warn("Metergraph authentication failed; capture disabled for this process");
        return;
      }
      if (response.status !== 202) throw new Error(`ingest returned ${response.status}`);
      this.backoffMs = 1_000;
      this.retryAt = 0;
    } catch {
      this.errors += 1;
      this.dropped += rows.length;
      this.retryAt = Date.now() + this.backoffMs;
      this.backoffMs = Math.min(this.backoffMs * 2, 60_000);
    }
  }

  async shutdown(): Promise<void> {
    if (this.timer) clearInterval(this.timer);
    await this.flush();
  }
}
